pkg/controller/ec2/vpcpeeringconnection/setup.go

```diff
@@ -0,0 +1,141 @@
+
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+// SetupVPCPeeringConnection adds a controller that reconciles VPCPeeringConnection.
+func SetupVPCPeeringConnection(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(svcapitypes.VPCPeeringConnectionGroupKind)
+	opts := []option{
+		func(e *external) {
+			c := &custom{client: e.client, kube: e.kube}
+			e.postObserve = c.postObserve
+			e.postCreate = c.postCreate
+			e.preCreate = preCreate
+			e.isUpToDate = c.isUpToDate
+			e.filterList = filterList
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&svcapitypes.VPCPeeringConnection{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.VPCPeeringConnectionGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type custom struct {
+	kube   client.Client
+	client svcsdkapi.EC2API
+}
+
+func filterList(cr *svcapitypes.VPCPeeringConnection, obj *svcsdk.DescribeVpcPeeringConnectionsOutput) *svcsdk.DescribeVpcPeeringConnectionsOutput {
+	connectionIdentifier := aws.String(meta.GetExternalName(cr))
+	resp := &svcsdk.DescribeVpcPeeringConnectionsOutput{}
+	for _, vpcPeeringConnection := range obj.VpcPeeringConnections {
+		if aws.StringValue(vpcPeeringConnection.VpcPeeringConnectionId) == aws.StringValue(connectionIdentifier) {
+			resp.VpcPeeringConnections = append(resp.VpcPeeringConnections, vpcPeeringConnection)
+			break
+		}
+	}
+	return resp
+}
+
+func (e *custom) postObserve(_ context.Context, cr *svcapitypes.VPCPeeringConnection, obj *svcsdk.DescribeVpcPeeringConnectionsOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	if awsclients.StringValue(obj.VpcPeeringConnections[0].Status.Code) == "pending-acceptance" && cr.Spec.ForProvider.AcceptRequest {
+		req := svcsdk.AcceptVpcPeeringConnectionInput{
+			VpcPeeringConnectionId: awsclients.String(*obj.VpcPeeringConnections[0].VpcPeeringConnectionId),
+		}
+		request, _ := e.client.AcceptVpcPeeringConnectionRequest(&req)
+		err := request.Send()
+		if err != nil {
+			return obs, err
+		}
+	}
+
+	available := setCondition(obj.VpcPeeringConnections[0].Status, cr)
+	if !available {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	return obs, nil
+}
+
+func setCondition(code *svcsdk.VpcPeeringConnectionStateReason, cr *svcapitypes.VPCPeeringConnection) bool {
+	switch aws.StringValue(code.Code) {
```

Comment:
We need to handle `VPCPeeringConnectionStateReasonCode_failed` here in the case that VPCs actually cannot be peered (bad vpcid, overlapping cidr, etc).

We do not want the controller to create new peering connection requests indefinitely.